ENH: Move identity to the ArrayMethod to allow customization

seberg opened the pull request:
This also fixes/changes that the identity value is defined by the
reduce dtype and not by the result dtype.  That does not make a
different in any sane use-case, but there is a theoretical change:

    arr = np.array([])
    out = np.array(None)  # object array
    np.add.reduce(arr, out=out, dtype=np.float64)

Where the output result was previously an integer 0, due to the
output being of `object` dtype, but now it is the correct float
due to the _operation_ being done in `float64`, so that the output
should be `np.zeros((), dtype=np.float64).astype(object)`.

---

This is a requirement for more advanced DTypes to pick a reasonable loop.  E.g. a if there was a reduce for a structured dtype, the structured dtypes default/initial element is unlikely to fill in from a 0.
Another example is the physical unit, that may reject casting `0` to `0 meters`, but needs the `0 meters` as identity.


(Note that depending on the order of merges, the wrapping array-method stuff requires a specialized version here to make my Unit dtype reductions work.)

seberg committed:
ENH: Move identity to the ArrayMethod to allow customization

This also fixes/changes that the identity value is defined by the
reduce dtype and not by the result dtype.  That does not make a
different in any sane use-case, but there is a theoretical change:

    arr = np.array([])
    out = np.array(None)  # object array
    np.add.reduce(arr, out=out, dtype=np.float64)

Where the output result was previously an integer 0, due to the
output being of `object` dtype, but now it is the correct float
due to the _operation_ being done in `float64`, so that the output
should be `np.zeros((), dtype=np.float64).astype(object)`.

seberg commented on numpy/core/src/umath/reduction.c:
```diff
@@ -198,38 +193,56 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
             NPY_ITER_DELAY_BUFALLOC |
             NPY_ITER_COPY_IF_OVERLAP;
+    if (!(identity_flags & NPY_METH_IS_REORDERABLE)) {
+        it_flags |= NPY_ITER_DONT_NEGATE_STRIDES;
```
I was expecting this change to speed up:
```
arr = np.ones(100000, 3)[::-1]
arr.sum()
```
turns out it doesn't... Something must not be optimized correctly in the iterator construction and now I am wondering if one of my old iterator-cleanup starts might include that fix...

mhvk replied:
Hmm, weird indeed. Still seems nice to add it here.

seberg committed:
ENH: Support identity-function in experimental DType API

Also add it to the wrapped array-method (ufunc) implementation
so that a Unit dtype can reasonably use an identity for reductions.

seberg commented:
This is almost ready, but it adds some (internal/experimental) API, and I am e.g. not sure about the names.  I do think this is the right place though (putting anything of this into the type resolver seems like too much).

One thing I now noticed is that we don't currently disable "reorderable" for object arrays or when `initial=None` is passed:
```
In [2]: arr = np.array([["a", "b"], ["c", "d"]], dtype=object)

In [3]: arr.sum()
Out[3]: 'abcd'

In [4]: arr.T.sum()
Out[4]: 'abcd'

In [5]: arr.T.sum(initial=None)
Out[5]: 'abcd'
```
The question is whether this should not reject reordering?  Further, if I allow stride negation, I suppose `arr = np.array(["a", "b", "c", "d"], dtype=object); arr[::-1].sum()` might still give `abcd` rather than `dcba`?  (I can't reproduce this right now.)

So that reorderable stuff, needs one more test or so, and the API some bike-shedding.  @mhvk if you have a bit of time, could you have a quick look?

seberg committed:
BUG: Fixup the default array-method, got the refcounting wrong...

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -313,12 +325,40 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
+                identity_flags & NPY_METH_ITEM_IS_IDENTITY ||
+                (empty_reduce && identity_flags & NPY_METH_ITEM_IS_DEFAULT))) {
+        /* Loop provided an identity or default value, assign to result. */
+        int ret = raw_array_assign_scalar(
```
It seems a bit confusing that this uses `raw_array_assign_scalar` rather than `PyArray_FillWithScalar`. Why are the built-in identities and user-provided ones treated differently? Is it still possible for the `get_identity` function to return a `PyObject` for filling with `dtype=object`? Part of the reason for wondering about always returning something reference counted elsewhere...

mhvk replied:
This has become irrelevant; I also see now that the tests done in `PyArray_FillWithScalar` are not needed any more. So, this can probably be resolved.

mhvk commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -284,6 +337,9 @@ fill_arraymethod_from_slots(
                 continue;
+            case NPY_METH_get_identity:
+                meth->get_identity = slot->pfunc;
+                continue;
```
Would be good to have tests for this!

seberg replied:
Yeah, I am using it in the unit dtype: https://github.com/seberg/unitdtype  So it is half tested, but not directly.  I guess I could add a test using the "scaled float"

mhvk replied:
Would still be a good idea...

mhvk commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -163,6 +166,55 @@ npy_default_get_strided_loop(
+        if (context->descriptors[0]->type_num != NPY_OBJECT) {
+            *flags |= NPY_METH_ITEM_IS_IDENTITY;
+        }
+        int res = PyArray_Pack(context->descriptors[0], item, identity_obj);
```
I wonder if it is really the best to do this packing inside this function, instead of trying to interpret the result. See main comment.

seberg replied:
I have a slight preference for using the raw data, also because I am still wondering if we might want dtypes without any scalar at all, and creating an array just to get the identity seems a bit heavy.

Alternatively, I could actually also `Pack` the `initial` provided by the user and use the `raw` version for both, although this version has the nice property that allowing arrays is changing one line to `CopyObject` :).

mhvk reviewed (commented):
Had a look but not quite complete, since I had some more general questions:
1. For object dtype, would it make sense to not have a default either? It just seems asking for trouble...
2. More on the implementation: should `reorderable` be separated from the identity? Conceptually, they are only somewhat related (e.g., in your string example, identity would be an empty string, but not reorderable). More generally, it would seem that the identify and whether a loop is reorderable can both depend on the dtype, but not necessarily the same way. Also, the code seems a bit contrived by combining the two.
Note that I realize you are just following what was there, but you are now on a path towards making it public API, so this seems the time to try to get it right...
3. Given above, perhaps an even more general question is whether the default identity and reorderability should be provided when adding a loop? That is the place where one truly knows what they are!
4. Or, even more general: is providing a function to retrieve the identity the best way, or might it be easier to provide the ability to have a custom function to do the initialization of the output? A bit like `__array_prepare__` used to be...

seberg commented:
> * For object dtype, would it make sense to not have a default either? It just seems asking for trouble...

Yeah, my worry is that in most cases this is completely fine for `object` arrays.  Arrays of `decimal`s can be summed and reordered just fine.  Removing the "default" would only change things for empty object arrays, the reorderable flag is tricky, because it currently disallows multi-dimensional arrays completely.

I suspect we _could_ remove the reorderable problem if we force C-order iteration (slowing things down, but keeping them working).  That way, for N-D without reorderable, we define the order rather than error out.

> * More on the implementation: should `reorderable` be separated from the identity? Conceptually, they are only somewhat related (e.g., in your string example, identity would be an empty string, but not reorderable). More generally, it would seem that the identify and whether a loop is reorderable can both depend on the dtype, but not necessarily the same way. Also, the code seems a bit contrived by combining the two.
>   Note that I realize you are just following what was there, but you are now on a path towards making it public API, so this seems the time to try to get it right...

I agree that they are two different things, but the reason I put them in the same callback is that they need to be queried at (roughly) the same time.

One approach could be to assume that `reorderable` is fully static.  Then we would set it on registration time and keep it fixed forever.
I thought it may be nice to avoid that (might give a bit more dynamism in the future), but I would be fine to drop that one (in principle it could still be added at some point, i.e. if reorderable is not set allow ufunc to query explicitly).

In that case I could pass in the information whether the reduction is empty (rather than pass out flags).

> * Given above, perhaps an even more general question is whether the default identity and reorderability should be provided when adding a loop? That is the place where one truly knows what they are!

Not sure I follow.  This moves the information to the `ArrayMethod`, i.e. to the loop.  But if the loop does _not_ provide the method, the default one is used (i.e. the one looking it up from the ufunc).

A possibility would be to use the default loop only for the current legacy loops and otherwise not provide a default at all (i.e. default to an exception).

> * Or, even more general: is providing a function to retrieve the identity the best way, or might it be easier to provide the ability to have a custom function to do the initialization of the output? A bit like `__array_prepare__` used to be...

Hmmm, I first thought this might be interesting for `gufunc`s, but now I am confused.  For reductions, we really only need a single element (unless we want to reduce a `gufunc` at some point).  `gufuncs` are normal ufuncs though, in the sense that they should not really have state (they might have working space or so, but `get_loop` should be able to deal with that).

One thing I don't like about `__array_prepare__` that I do not want to pass in the full arrays.  The ufunc does not need to know that it operates on an N-dimensional NumPy array currently.  It provides a single element for the `identity`  (in principle, this would be the core dimensions of a gufunc, but I am not sure what this means?).

seberg commented on numpy/core/src/umath/reduction.c:
```diff
@@ -198,38 +193,54 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
+    char *identity_buf = NULL;
+    if (initial == NULL) {
+        /* Always init buffer (only necessary if it holds references) */
+        identity_buf = PyMem_Calloc(1, op_dtypes[0]->elsize);
```
Need to NULL it if the dtype requires initialization (holds objects effectively), if that is not the case, it doesn't matter.

seberg commented on numpy/core/src/umath/reduction.c:
```diff
@@ -313,12 +325,40 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
+        }
+    }
+    else if (identity_buf != NULL && (  /* cannot fill for `initial=None` */
+                identity_flags & NPY_METH_ITEM_IS_IDENTITY ||
```
Good idea, then the method would also be `get_initial` rather than `identity`.  Will rename (not right now).

seberg commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -248,6 +300,7 @@ fill_arraymethod_from_slots(
     /* Set the defaults */
     meth->get_strided_loop = &npy_default_get_strided_loop;
     meth->resolve_descriptors = &default_resolve_descriptors;
+    meth->get_identity = default_get_identity_function;
```
Yup about the &, I prefer it (both work, but without seems implicit to me).
The naming actually has a purpose to the madness, the `npy_` one is exported for use elsewhere.

seberg committed:
DOC,MAINT: Address "simple" review comments by Marten

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -313,12 +325,40 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
+    if (initial != NULL && initial != Py_None) {
+        /*
+         * User provided an `initial` value and it is not `None`.
+         * NOTE: It may make sense to accept array-valued `initial`,
```
Note the suggestion of an array-valued `initial` for `bincount` at https://github.com/numpy/numpy/issues/22471#issuecomment-1288176870

seberg replied:
Yeah, but it isn't important for this PR (but I could prioritize it for a follow-up!).  I am planning on rewording things to use `initial`, and then say whether it is "identity" or "default".  And also allow the flexibility of using identity (only when reduction is non-empty) rather than default (only when it is empty).
The example being `-inf` for `max()`, when we want `max([])` to error (although that is likely not what we want!).

Just busy rewording, the main thing here (from my point of view) is still whether "reorderable" should just be considered "static" (mainly to simplify the `get_identity()`/`get_reduction_initial()` function), although the dynamic makes it a bit easier to fall back to the ufunc!

mhvk replied:
OK, sounds good. It does seem to me that `reorderable` can be static for any given loop -- looking at this briefly again, the combination with the identify continues to feel odd.

seberg replied:
Thanks.  OK, I got some more refactoring to do here, then will see if there is a good way to disentangle it and make the info static.
I do think it is fathomable that it isn't static (i.e. dtype instance dependent), but the main thing will be if it seems too clunky w.r.t. to falling back to the current mechanism.

seberg committed:
MAINT: Rename arraymethod reduction identity getter

Need to look into whether to cut out the dynamic discovery of
reorderability though.

seberg committed:
MAINT: Fixup names, and add workaround lost in rebase

seberg committed:
MAINT: Rework things to make reorderability static and default to no initial value

seberg committed:
BUG: Fixup copying the wrong size in the cached path mostly

seberg committed:
API: Bumpy experimental dtype api version

Its a new version (it changed ;)), plus I took the liberty to move
things around a bit ABI wise.

seberg committed:
TST: Fix lint and ensure we have tests for empty arrays and reduce

seberg commented:
OK, took me a bit to settle on things.  But I think I am happy with it now:
* There is a single flag `NPY_METH_IS_REORDERABLE`.  We could use this more generally, but for it is only used for reductions.  (This removes the ability to decide on reorderability based on the actual dtype instance, that is probably OK and there should be way to add it if it turns out very useful.)
* The "legacy" ArrayMethods, query the ufunc on initialization for this flag, and for wether an identity should exist.
* I took the liberty of caching the identity for numeric loops.  This shaves off ~5% of overhead, but it is very straight forward, so I like it.
* Any non-legacy loop will default to no identity and not reorderable.

We pass in `reduction_is_empty` now, the implementor has to return 1 for "initial value exists" and 0 if not (-1 on critical error).

seberg committed:
MAINT: Remove unneeded label

mhvk commented on numpy/core/src/umath/legacy_array_method.c:
```diff
@@ -273,6 +359,27 @@ PyArray_NewLegacyWrappingArrayMethod(PyUFuncObject *ufunc,
+            flags |= NPY_METH_IS_REORDERABLE;
+        }
+        if (identity_obj != Py_None) {
+            /* NOTE: We defer, just in case it fails in weird cases: */
```
I don't understand this comment!

seberg replied:
Hehe, this calls `PyArray_Pack(identity_value)`.  If you wrote a weird user dtype, maybe that actually fails for certain values (say -1).  In that case, I don't want to fail at ufunc initialization time, since maybe it never worked, but nobody would call `ufunc.reduce` later.

mhvk replied:
OK, that makes sense!  Maybe just update the comment with what you wrote above!

seberg replied:
I decided to just delete it here and expand the other one (changes pending).

For NumPy, it is convenient right now to write generic code.  If we were not doing that, the functions would just use static things anway, i.e. be written as (very roughly):
```
static int
copy_initial_data<T>() {
    *<T *>initial = 0;
    return 0;
}
```
so yes, I didn't mean to make this available (at least for now).

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -297,8 +288,50 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
+    }
+    else {
+        /* Not all functions will need initialization, but init always: */
+        initial_buf = PyMem_Calloc(1, op_dtypes[0]->elsize);
```
Hmm, this makes me wonder if the initial value getter should just return a pointer to the initial value. That saves the `memcpy` too.

seberg replied:
Yeah, but in that set-up who free's the data?  It works beautifully if that data would be static (as is usually the case), but if we need to copy the data from something more dynamic (say an initial value for an arbitrary precision number of non-fixed size; or a structured dtype)...

Then cleanup of that data would be impossible if desired, it would have to be attached to the dtype instance or so (maybe that is OK, but not sure).

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -395,6 +447,10 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
     return result;
 
 fail:
+    if (initial_buf != NULL && PyDataType_REFCHK(op_dtypes[0])) {
```
Would it be possible to add a test for this?

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2067,12 +2067,16 @@ _get_coredim_sizes(PyUFuncObject *ufunc, PyArrayObject **op,
-static PyObject *
-_get_identity(PyUFuncObject *ufunc, npy_bool *reorderable) {
+NPY_NO_EXPORT PyObject *
+PyUFunc_GetIdentity(PyUFuncObject *ufunc, npy_bool *reorderable)
```
Should this be `_GetLegacyIdentity`? If indeed only used for old-style ufuncs, perhaps the function can be moved to `legacy_array_method`?

Or is this also meant to be used for new ufuncs? If so, would it be better to split out the `reorderable` and `identity` parts here too? (and adjust `ufunc_get_identity` accordingly).

seberg replied:
Honestly, I still need to figure that out (thus no change yet).  The problem is, we still have `ufunc.identity` and maybe someone actually uses that?  But if we want downstream to actually use this (rather than always implement the an initial getter), then we should probably add it as public API and give it some clearly name like `GetUFuncDefaultIdentity`.

mhvk replied:
Indeed, we need to support `.identity` since someone might well use that. Looking at this again, I now agree it is best to just keep the code as close as possible as it is here, since we need the code as is in `legacy_array_method` and this works fine for getting the `.identity` attribute. Really what tripped me up here was the name change, both the name itself and the impression from the `PyUFunc` start that it would (eventually) be public. So, I guess the first question is whether we even need to change the name at all? I guess the reason for that is that the function is now used inside `legacy_array_method`, which makes sense. And I guess I forgot that `PyUfunc` just indicates "used elsewhere in the numpy C code"; it does have an `NPY_NO_EXPORT` after all!
Anyway, bottom line is that I think all we really need is a clearer name, and I like your `PyUFunc_GetDefaultIdentity` (removing the extra `UFunc` you had in your comment).

We can then leave open for later a possible public API (where perhaps one might want to get the identity for a given loop or so).

seberg replied:
OK, renamed to `PyUFunc_GetDefaultIdentity`... maybe it would make sense to have a pattern of using NPY for private and PyArray style for public. But I suspect we break it in both ways too often.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -6459,7 +6415,7 @@ static PyObject *
 {
     npy_bool reorderable;
-    return _get_identity(ufunc, &reorderable);
+    return PyUFunc_GetIdentity(ufunc, &reorderable);
```
I think it would be nice to change this too, to not need the `reorderable` any more.

seberg replied:
Happy to change this, but as it is, if someone were to think of adding a new "default", they at least see that the two are currently linked?

mhvk commented on numpy/core/src/umath/wrapping_array_method.c:
```diff
@@ -177,6 +177,39 @@ wrapping_method_get_loop(
+ * We assume again that translating the descriptors is quick.
+ */
+static int
+wrapping_method_get_identity_function(
```
Would it be possible to have a test for this?

mhvk commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -1617,6 +1617,18 @@ def test_identityless_reduction_noncontig_unaligned(self):
+        out = np.array(None, dtype=object)
+        # When the loop is float but `out` is object this does not happen:
+        np.add.reduce([], out=out, dtype=np.float64)
+        assert type(out[()]) is not int
```
Can we also test what it should be? (`float`?)

mhvk commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -1646,6 +1661,18 @@ def test_initial_reduction(self):
+        # Test that an empty reduction fails also if the result is empty
+        arr = np.zeros((0, 0, 5))
+        with pytest.raises(ValueError):
+            np.true_divide.reduce(arr, axis=1)
```
Arguably overkill, but perhaps also add a line where one does give an `initial`?

mhvk commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -2546,7 +2573,8 @@ def test_reduce_casterrors(offset):
         # This is an unsafe cast, but we currently always allow that.
         # Note that the double loop is picked, but the cast fails.
-        np.add.reduce(arr, dtype=np.intp, out=out)
+        # `initial=None` disables the use of an identity here.
```
Is this just to ensure refcounts on it do not contribute?

seberg replied:
Expanding the comment, the point is that the cast error may happen while copying the "first values", but that only happens when there is no initial.

mhvk reviewed (commented):
This seems quite nice! And array method now quite logical. My main question is whether the caching could not be made more general by having it be a pointer to data (which also avoids copying it).

Another is whether one can push the separation of `reorderable` and `identity` a bit further.

mhvk commented on numpy/core/src/multiarray/array_method.h:
```diff
@@ -203,6 +226,8 @@ typedef struct PyArrayMethodObject_tag {
     translate_given_descrs_func *translate_given_descrs;
     translate_loop_descrs_func *translate_loop_descrs;
+    /* Chunk used by the legacy fallback arraymethod mainly */
+    char initial[sizeof(npy_clongdouble)];  /* initial value storage */
```
True that it is not quite general - still, since the allocation is done elsewhere and data is copied from this location, I'm not sure there is a net saving in complexity in having what seems a rather odd difficult to explain piece of `PyArrayMethodObject_tag` -- it just doesn't feel properly extensible/re-usable, unlike the rest.

mhvk replied:
This is the one part that still feels super clunky. I complain in part because the rest feels really nice and logical yet here were suddenly have some opaque blob of bytes! 

Looking at the code a bit more, a pointer to some memory allocated on first call would not be difficult right? You'd just have to be careful to deallocate it in `arraymethod_dealloc`, like now done with `name`.

seberg replied:
No, it would be easy to do such a pointer.  Maybe our difference is that I still consider `PyArrayMethodObject` to be fully opaque and private to NumPy?

mhvk replied:
I guess the basic issue remains that having a chunk of random bytes feels clunky, but you are right that to me the struct really seems something that could become public (or, if not, a really useful thing for developers to look at), so one should keep it neat. And if it becomes public, having a pointer seems far more useful. Probably even better if it is a pointer to something well-defined like an array scalar.

seberg replied:
Hmmm, it is filled always with the same value and when the GIL is held, so I think it is fine.  Will adept the comment to "only" not "mainly".  Because you are right: reusing it outside of the "legacy" path might be prone to threadsafety issues.

seberg commented on numpy/core/include/numpy/experimental_dtype_api.h:
```diff
@@ -321,6 +327,29 @@ typedef int (PyArrayMethod_StridedLoop)(PyArrayMethod_Context *context,
+ *
+ * @param context The arraymethod context, mainly to access the descriptors.
+ * @param initial Pointer to initial data to be filled (if possible)
+ * @param reduction_is_empty Whether the reduction is empty, when it is the
```
No opinion on it, so swapping it :).

seberg committed:
MAINT: Address many of Marten's comments

seberg committed:
TST: Add test for wrapped ufuncs and reductions via ScaledFloat

seberg committed:
MAINT: Rename PyUFunc_GetIDentity to PyUFunc_GetDefaultIdentity

seberg commented:
Added another test using the ScaledFloat stuff, that should also cover a whole lot more really.  Will check on coverage tomorrow.  Hopefully, I addressed most issues.  I did not do anything about the way I cache things:  Yes, this is not public API, but simply something we do within NumPy (and maybe that is OK, since NumPy numbers are a bit special in that future dtypes would mostly use a single implementation per DType).

Thanks for the thorough reviews, Marten!

mhvk commented on numpy/core/src/umath/legacy_array_method.c:
```diff
@@ -234,6 +237,94 @@ get_wrapped_legacy_ufunc_loop(PyArrayMethod_Context *context,
+
+    if (PyTypeNum_ISNUMBER(context->descriptors[0]->type_num)) {
+        /* For numbers we can cache to avoid going via Python ints */
+        memcpy(context->method->initial, initial, context->descriptors[0]->elsize);
```
So, my sense still would be to just allocate the right number of bytes here, copy over initial to those, and store the pointer (maybe `context->method->cached_initial`?).

Alternatively, making it all even less opaque, I guess one could repack `initial` here as a array or numpy scalar (which now has the right `dtype`) and cache that? Then, the replacement function could simply return the pointer to the data contained in that cached array and in the deallocation, all one would need is a `Py_XDECREF(meth->cached_initial)`.

seberg replied:
Hmmm, my sense is still that this is simpler as long as it is just a small, private, optimization for now.
To me allocating it feels like we should provide infrastructure so that users can make use of that machinery.  And I am not really motivated to aim for such a mechanism at this point (seems like little gain for additional complexity).

mhvk replied:
If it was hidden in an obviously private place, not in a `struct` that feels like something that should become public, I'd agree!

seberg replied:
Maybe we need a third opinion :).  Yes we may make (part of) that struct public, and then we would have to hide this away.
But I like keeping the cache and I don't like adding the complexity of an allocation at this time...

mhvk replied:
Sure, this shouldn't hold up what is a really nice PR -- maybe if someone else can have a quick look and see whether it is worth the effort doing an allocation, that would indeed be best. In the end it is really a "does this pass the smell test" matter and my sense of smell is not necessarily the best, especially for C.

mhvk reviewed (commented):
@seberg - last try to argue for a pointer and allocating memory rather than having a blob! See in-line for a bit more concrete suggestions

seberg commented:
Would be nice to move this one forward soon (and the [decref PR](gh-22924), although unrelated).  Will mark it again for triage-review, just to nudge it :).

mhvk commented:
This remains really nice and should go in, though it would be good if someone else could review, and maybe decide on the only bit that I didn't quite like (see https://github.com/numpy/numpy/pull/20970#discussion_r1024542190 and rest of chain of comments).

ngoldbaum commented on numpy/core/include/numpy/experimental_dtype_api.h:
```diff
@@ -321,6 +327,28 @@ typedef int (PyArrayMethod_StridedLoop)(PyArrayMethod_Context *context,
+ * Query an ArrayMethod for the initial value for use in reduction.
+ *
+ * @param context The arraymethod context, mainly to access the descriptors.
+ * @param reduction_is_empty Whether the reduction is empty, when it is the
+ *     default value is required, otherwise an identity value to start the
+ *     the reduction.  These might differ, examples:
+ *     - `0.0` as default for `sum([])`.  But `-0.0` would be the correct
+ *       identity as it preserves the sign for `sum([-0.0])`.
+ *     - We use no identity for object, but `0` and `1` for sum and prod.
+ *     - `-inf` or `INT_MIN` for `max` is an identity, but at least `INT_MIN`
+ *       not a good *default* when there are no items.
```
I had to squint at this a little bit to understand what was going on, so I'd suggest the following reword to make things clearer:

```suggestion
 * @param reduction_is_empty Whether the reduction is empty. When it is, the
 *     default value for the identity might differ, for example:
 *     - `0.0` is the default for `sum([])`.  But `-0.0` is the correct
 *       identity otherwise as it preserves the sign for `sum([-0.0])`.
 *     - We use no identity for object, but `0` and `1` for sum and prod.
 *     - `-inf` or `INT_MIN` for `max` is an identity, but at least `INT_MIN`
 *       not a good *default* when there are no items.
```

If any of my edits change the meaning of what you were going for please discard, but maybe try re-wording because I'm still confused about the second example at least. Also this documentation is copy/pasted in `array_method.h`, so make sure to keep that in sync.

seberg replied:
Thanks, the object case should be: [...] , but use the defaults `0` and `1` for sum and prod" with no elements.  (will change)

ngoldbaum commented on numpy/core/include/numpy/experimental_dtype_api.h:
```diff
@@ -321,6 +327,28 @@ typedef int (PyArrayMethod_StridedLoop)(PyArrayMethod_Context *context,
+ *     may call this even when not strictly necessary.
+ */
+ #define NPY_METH_get_reduction_initial 4
+typedef int (get_reduction_intial_function)(
```
```suggestion
typedef int (get_reduction_initial_function)(
```

ngoldbaum replied:
The same typo occurs in the typedef in `array_method.h`

seberg replied:
Yeah, numbered it the same way as the struct now, that makes most sense if I change it already...

ngoldbaum commented on numpy/core/include/numpy/experimental_dtype_api.h:
```diff
@@ -321,6 +327,28 @@ typedef int (PyArrayMethod_StridedLoop)(PyArrayMethod_Context *context,
+ *     successfully filled.  Errors must not be given where 0 is correct, NumPy
+ *     may call this even when not strictly necessary.
+ */
+ #define NPY_METH_get_reduction_initial 4
```
It looks like you updated the values of the other `NPY_METH` constants in array_method.h, but not here.

Maybe it would help to catch stuff like this if the scaled float dtype used this header.

ngoldbaum reviewed (commented):
Saw you wanted some eyes on this one along with #22924 so I figured I'd give it a once over. I don't have any substantive comments about the implementation but I did have some comments about documentation and I think there might be a bug in the changes to the experimental dtype header.

seberg committed:
MAINT: Address review comments by Nathan

mattip commented on numpy/core/src/umath/ufunc_object.h:
```diff
@@ -12,6 +12,9 @@ ufunc_seterr(PyObject *NPY_UNUSED(dummy), PyObject *args);
+NPY_NO_EXPORT PyObject *
+PyUFunc_GetDefaultIdentity(PyUFuncObject *ufunc, npy_bool *reorderable);
+
 /* strings from umathmodule.c that are interned on umath import */
```
Is this "get reduction default" for iterating over empty arrays or "get reduction initial"?

seberg replied:
Neither quite unfortunately :(.  Because it is the initial always *except* for object arrays.

mattip reviewed (commented):
Looks good. Mainly some nits about duplicated code and names. As for the static field in the struct, I think it is OK for now, but should be refactored in the future.

seberg commented on numpy/core/src/multiarray/array_method.h:
```diff
@@ -104,6 +104,28 @@ typedef int (get_loop_function)(
+ *
+ * @param context The arraymethod context, mainly to access the descriptors.
+ * @param reduction_is_empty Whether the reduction is empty. When it is, the
+ *     default value for the identity might differ, for example:
```
```suggestion
 *     initial value for the identity might differ, for example:
```
The function returns the "initial" value.  The term "default" (for the value returned by the empty reduction) should only be used in the bullet points later.  

I don't hate the term "default", but we don't use it anywhere.  So I think Matti is right and I should rephrase this to, at least mainly, just not use it.

seberg committed:
MAINT: Prepend scaled float test statics with `sfloat_`

As requested by Matti in review.

seberg commented on numpy/core/src/umath/_scaled_float_dtype.c:
```diff
@@ -645,20 +653,74 @@ add_sfloats_resolve_descriptors(
+
+static int
+add_loop(const char *ufunc_name,
+        PyArray_DTypeMeta *dtypes[3], PyObject *meth_or_promoter)
```
Prepending all in this file with `sfloat_` for now, indeed a bit clearer to not share names even if static.

seberg commented on numpy/core/src/umath/reduction.c:
```diff
@@ -297,8 +288,50 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
 
+    /*
+     * Get the initial value (if it may exists).  If the iteration is empty
+     * then we assume the reduction is (in the other case, we do not need
```
Grammatically and logically correct, but super confusing also for me reading it :).

seberg commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2068,12 +2068,16 @@ _get_coredim_sizes(PyUFuncObject *ufunc, PyArrayObject **op,
+ * Returns a new reference to the ufunc identity.  Note that this identity
+ * is only a default identity value stored on the ufunc, since the invidiual
+ * ufunc loop (ArrayMethod) is queried for the actual identity.
+ *
```
Yes, since this is what is stored as `ufunc.identity`, it seems easier to keep it as "identity" here.

seberg committed:
DOC: Adept code comments for clarity based on code review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seberg commented:
Thanks for the review, I have addressed the comments except those pertaining to the duplications/fixups in `experimental_dtype_api.h` (and maybe related).

I don't disagree with them, but didn't have a great idea and also think it may be better to clean up as a follow-up.  Ping also @ngoldbaum, I guess the solution is to create a new public header that is included both by the internal one and external ones (which includes all the things that are truly identical)

mattip commented:
> I guess the solution is to create a new public header that is included both by the internal one and external one

+1

mattip commented:
extra points if the external one does not expose the actual fields of the structures.

mattip commented:
I think the circleCI failure is due to a merge problem with older python still being used in the build

seberg commented:
> I think the circleCI failure is due to a merge problem with older python still being used in the build

Yeah, I could rebase to fix it (although the built is currently a bit broken due to gh-23034).

mhvk commented:
Happy for this to go in with @mattip's comments addressed. Perhaps particularly in making the names really reflect what is legacy and what is not (no hack for the new stuff!).

seberg committed:
MAINT: Rename `initial` to `legacy_initial` in ArrayMethod

It should only be used by the legacy method, so also reflect that
in the field name.

seberg commented:
Ah, I hand't renamed the field, so done that now.  I will look into the header changes as a followup.  (Also rebased, for the sake of circle CI mainly).

mattip commented on numpy/core/include/numpy/experimental_dtype_api.h:
```diff
@@ -321,6 +327,28 @@ typedef int (PyArrayMethod_StridedLoop)(PyArrayMethod_Context *context,
+ *     successfully filled.  Errors must not be given where 0 is correct, NumPy
+ *     may call this even when not strictly necessary.
+ */
+#define NPY_METH_get_reduction_initial 3
```
Is this here on purpose or can it move up to the other defines?

seberg replied:
I have it grouped the corresponding typedef and there are 2 more of these that are "stray" in that sense.  I am happy to put them together but because of that didn't.

mattip reviewed (approved):
LGTM, thanks. One small nit, but that can be part of the follow on PR to deduplicate the headers.

ngoldbaum commented:
Since this updates the experimental dtype API and we're planning to also add a new hook for a clear function in a followup to #22924, it might make sense to batch those two API breaks together. That said, I'm happy to clean up the dtypes in `numpy-user-dtypes` twice if we'd rather have two API breaks.

I agree that it would be best if the internal headers and external headers could share common definitions. That will avoid repeats of #21979 and the inconsistency between the headers I caught in this PR in the future.

seberg commented:
Yeah, the cleanup PR doesn't include exposing the new API yet.  If the other PR can move forward in the next days, we can effectively squash the changes.  But, I would prefer uncoupling the two things:  Lets just get this over with, and deal with whatever small fallout exists for the experimental DTypes.

mattip commented:
Thanks @seberg and @mhvk
